### ISSUE TITLE
Stuck with FileProcessLock lock entries in database_fileprocesslock

### ISSUE THREAD

**Issue by usanchez** [onnela-lab/beiwe-backend#72]
So, I got this message on my mail:

> DATA PROCESSING OVERLOADED, CHECK SERVER.
> Data processing has overlapped with a prior data index run that started more than 15 minutes ago.
> That prior run has been going for 17 hour(s), 14 minute(s)

I checked the code and I realised that this error is provoked by a file being locked for processing, and apparently not being unlocked. I checked the database and I saw that the file being locked was part of a test study that didn't matter too much, and looking at what FileProcessLock.unlock() function does, I simply deleted this entry from the database.

Then, a new entry is generated in _database_fileprocesslock_, but the problem is, that the id of this file does not correspond to any file in the _filetoprocess_ table. After I remove that entry, a new one is generated, with the id of the file increased by one, and again, this file does not exist in _database\_filetoprocess_ table.

So I have quite a lot of files pending to be processed. What would be the best way to fix this issue so that the files start to be processed correctly again? How can fix and reset this loop? I've looked at other issues, but I don't know what and how I should run anything to solve this problem.

I think that the version of my platform is from May

Thank you very much.

Best regards,

Unai.

**Comment by biblicabeebli** [comment]
I've seen this, on my commute, will check back soon.

please send me a stack trace (not of the processing overload, the
associated error from within the data processing code.)

Try this: update to the newest version of the code (both on elastic
beanstalk servers and the data processing server(s)) on the production
branch.

The FileProcessingLock database table is the entity that controls whether
the manager server dispatches tasks to the processing servers.  The
.unlock() function clears the lock.  The table does not contain files.

File processing runs that generate errors will force data processing to
fail in a way that blocks the next attempts until the error is checked.

Other known failure modes include one user constantly uploading files ,
because then processing will "never finish" for that user, blocking the
next processing.

Could you please restate the action that you manually did?  It is slightly
ambiguous which tables you did each operation on.

**Comment by usanchez** [comment]
Thank you very much for answering. I'll try to go point by point.
> I've seen this, on my commute, will check back soon. please send me a stack trace (not of the processing overload, the associated error from within the data processing code.)
Sorry, but I don't know how to check this information that you request. Are you referring to the output of the method _report_file_processing_locked_and_exit_? Or should I look into the _error\_log_ files? 

> Try this: update to the newest version of the code (both on elastic beanstalk servers and the data processing server(s)) on the production branch.
Thanks. I'll start with this tomorrow (12h from now for me).

> The FileProcessingLock database table is the entity that controls whether the manager server dispatches tasks to the processing servers. The .unlock() function clears the lock. The table does not contain files.
Yeah, that was what I understood.

> File processing runs that generate errors will force data processing to fail in a way that blocks the next attempts until the error is checked. Other known failure modes include one user constantly uploading files , because then processing will "never finish" for that user, blocking the next processing.
I think that users are sending their files every 15 mins. Is there a way to check if any user is constantly uploading files?

> Could you please restate the action that you manually did? It is slightly ambiguous which tables you did each operation on.

1. I manually deleted the only entry that appeared in _database_fileprocesslock_.
2. Checked _database_fileprocesslock_. Saw that the entry reappeared.
3. Deleted manually the entry in  _database_filetoprocess_ that was causing the problem.
4. Deleted the entry that appeared in _database_fileprocesslock_.
5. A new entry appeared in _database_fileprocesslock_, with an ID that does not exist in  _database_filetoprocess_. I deleted the entry from _database_fileprocesslock_.
6. A new entry appeared in _database_fileprocesslock_, with an ID (ID from _step 5_ + 1) that does not exist in _database_filetoprocess_. I didn't do anything else.

I did not run any code whatsoever.

Hope that I explained well myself! Thank you very much!

**Comment by biblicabeebli** [comment]
Before anything else, you need to check if there is a long running data processing event occuring.  I recommend `htop`, but `ps -ef | grep python` may be sufficient for you to work it out.  Kill that process if it is running, but let me know the stats that it has (memory usage, cpu usage, etc.)


Stack trace - ah, you got this as an email, not as an error report; no stack trace.

Do this only after upgrading the codebase, you will probably need to run a pip install of the Requirements.txt file.  (possibly with `sudo`, `pip install -r Requirements.txt`)

You will then need to run this command, it may take a little while (shouldn't be more than a couple minutes and only if you are super out of date):
```
python manage.py migrate
```
run `crontab -e`, comment out the cron jobs.  this will stop the failure from occurring constantly and let you debug.  Don't forget to uncomment when you are done.

run `python manage.py shell_plus` - this will open a (fantastic) ipython shell and import a bunch of stuff so that you can interact with the code and the database.  It has tab completion!

I'm going to have you manually unlock and then run processing.  Send me the whole output you get.  If you want to you can sanitize out username strings, but those are already randomized so your call on that one:
```
FileProcessLock.unlock()
from libs.file_processing import process_file_chunks
process_file_chunks()
```

You will get a stack trace, possibly the "everything went fine" stack trace.  Copy-paste it in your response.


=====

There is a tool I created to watch active user uploads.... but that's on the python 3 branch....

you can enter this into the ipython shell, it will display a moderately readable list of the most recent 100 files recently uploaded and waiting to be processed.
```
list(FileToProcess.objects.order_by("-created_on")[:100].values())
```
You can also run this to get the last 100 files uploaded (may be sloooooow, may be empty if you are super out of date):
```
list(UploadTracking.objects.order_by("-created_on")[:100].values())
```

**Comment by usanchez** [comment]
Thank you very much for your help. The answers are again after your quotes.

> Before anything else, you need to check if there is a long running data processing event occuring. I recommend `htop`, but `ps -ef | grep python` may be sufficient for you to work it out. Kill that process if it is running, but let me know the stats that it has (memory usage, cpu usage, etc.)

Here is the result to htop:
![imagen](https://user-images.githubusercontent.com/6339300/68766933-1a688900-0620-11ea-9a28-df02cd39052c.png)

Is the process that I need to kill the process 4753?

> Stack trace - ah, you got this as an email, not as an error report; no stack trace.

Yeah, I need to set up a Sentry server to gather these stack traces.

> Do this only after upgrading the codebase, you will probably need to run a pip install of the Requirements.txt file.  (possibly with `sudo`, `pip install -r Requirements.txt`)

Should I upgrade the code and then just deploy it with eb deploy?  To continue with the next steps? Or just upgrade it locally?

> You will then need to run this command, it may take a little while (shouldn't be more than a couple minutes and only if you are super out of date):
> 
> ```
> python manage.py migrate
> ```
> 
> run `crontab -e`, comment out the cron jobs.  this will stop the failure from occurring constantly and let you debug.  Don't forget to uncomment when you are done.
> 
> run `python manage.py shell_plus` - this will open a (fantastic) ipython shell and import a bunch of stuff so that you can interact with the code and the database.  It has tab completion!
> 
> I'm going to have you manually unlock and then run processing.  Send me the whole output you get. If you want to you can sanitize out username strings, but those are already randomized so your call on that one:

From my local computer or from any of the servers?

Thank you very much and sorry for the inconvinience.

**Comment by usanchez** [comment]
I tried to do these steps:
> run `python manage.py shell_plus` - this will open a (fantastic) ipython shell and import a bunch of stuff so that you can interact with the code and the database.  It has tab completion!
> 
> I'm going to have you manually unlock and then run processing.  Send me the whole output you get. If you want to you can sanitize out username strings, but those are already randomized so your call on that one:
> 
> ```
> FileProcessLock.unlock()
> from libs.file_processing import process_file_chunks
> process_file_chunks()
> ```
> 
> You will get a stack trace, possibly the "everything went fine" stack trace. Copy-paste it in your response.
> 

So I run these lines with the amazing shell :D and I got this:
```
ubuntu@ubuntu14vm:~/humanai-backend$ python manage.py shell_plus
# Shell Plus Model Imports
from database.data_access_models import ChunkRegistry, FileProcessLock, FileToProcess, PipelineRegistry, PipelineUpload, PipelineUploadTags
from database.profiling_models import DecryptionKeyError, EncryptionErrorMetadata, LineEncryptionError, UploadTracking
from database.study_models import DashboardColorSetting, DashboardGradient, DashboardInflection, DeviceSettings, Study, StudyField, Survey, SurveyArchive
from database.user_models import Participant, ParticipantFieldValue, Researcher, StudyRelation
# Shell Plus Django Imports
from django.urls import reverse
from django.utils import timezone
from django.conf import settings
from django.core.cache import cache
from django.contrib.auth import get_user_model
from django.db.models import Avg, Case, Count, F, Max, Min, Prefetch, Q, Sum, When
from django.db import transaction
# Shell Plus User Imports
from libs.s3 import s3_list_files
from pprint import pprint
from datetime import date, datetime, timedelta
Python 2.7.13 (default, Jan 17 2019, 11:32:33) 
Type "copyright", "credits" or "license" for more information.

IPython 5.5.0 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: FileProcessLock.unlock()

Not-critical exception: Settings already configured.
In [2]: from libs.file_processing import process_file_chunks

In [3]: process_file_chunks()
---------------------------------------------------------------------------
ProcessingOverlapError                    Traceback (most recent call last)
<ipython-input-3-fba1ac43c870> in <module>()
----> 1 process_file_chunks()

/home/ubuntu/humanai-backend/libs/file_processing.py in process_file_chunks()
     37     error_handler = ErrorHandler()
     38     if FileProcessLock.islocked():
---> 39         raise ProcessingOverlapError("Data processing overlapped with a previous data indexing run.")
     40     FileProcessLock.lock()
     41 

ProcessingOverlapError: Data processing overlapped with a previous data indexing run.
```
I don't know if there is much information here...

Thank you very much!

**Comment by biblicabeebli** [comment]
1) tasks to kill: yes, those chronic five_minutes processes are the ones to kill.

2) Every action I was talking about were for a manager server.

You will also have to upgrade the worker server, and the elastic beanstalk server.  I assume you are capable of dealing with git issues and making sure all three are on the same commit/branch (`production`, there is also a `python2-legacy` branch that you will need to swap to if you do not upgrade your server stack when the python 3 update is finalized.)

Upgrading worker server: same as on a manager, but you also have to run `processing-stop` and `processing-start` commands. (you can look it htop for "celery" workers, hit shift-H to hide irrelevant threads, and T to put it into a tree/hierarchical view.  You can also click on CPU%, MEM% etc [Yes. Click. HTOP rocks] to sort the column, but tree view + sorting is limited.)\

On the EB servers you will have to do a deploy operation from your local machine.

3) That stack trace:
That is the overlapping run stacktrace, that is the one that triggers the email you are receiving.  I can only assume that the five-minutely task ran and failed between your unlock command and your calling `process_file_chunks`.  Kill those processes, disable cron job, try again.

**Comment by usanchez** [comment]
Here we go!

>     1. tasks to kill: yes, those chronic five_minutes processes are the ones to kill.
All killed, no survivors.

>     2. Every action I was talking about were for a manager server.

Thanks!

> You will also have to upgrade the worker server, and the elastic beanstalk server. I assume you are capable of dealing with git issues and making sure all three are on the same commit/branch (`production`, there is also a `python2-legacy` branch that you will need to swap to if you do not upgrade your server stack when the python 3 update is finalized.)

EB, processing manager and processing server upgraded! Congratulations on your progress by the way, it must have been tough to make the change from 2 to 3, and you are already in the Done v2 phase! :D

> Upgrading worker server: same as on a manager, but you also have to run `processing-stop` and `processing-start` commands.

Had to run `sudo processing-start` for `supervisor` to start. I couldn't do it without `sudo`, hope that I'm not running it incorrectly.

> (you can look it htop for "celery" workers, hit shift-H to hide irrelevant threads, and T to put it into a tree/hierarchical view. **You can also click on CPU%, MEM% etc [Yes. Click. HTOP rocks]** to sort the column, but tree view + sorting is limited.)\

Just did that 😮 

> On the EB servers you will have to do a deploy operation from your local machine.

Already deployed.
 
>     1. That stack trace:
>        That is the overlapping run stacktrace, that is the one that triggers the email you are receiving.  I can only assume that the five-minutely task ran and failed between your unlock command and your calling `process_file_chunks`.  Kill those processes, disable cron job, try again.

Tried again. Is still running (it has been like this for 4 hours or so), but seems like everything is running ok. `python manage.py shell_plus` is amazing by the way.

**Comment by biblicabeebli** [comment]
(ipython, shell plus, and htop make ssh server management tolerable.)

Thanks on the python 3, there were some obnoxious bits tracking down every
single string operation involved in encryption/decryption and updating that
logic.

`sudo processing start` should be irrelevant.

Okay, sounds like you are uploading more data than you can process,
hooray!  Do some calls of `FileToProcess.objects.count()`.  If your
python-fu is up to it make a loop that prints that value every..... 10
seconds?

Actually, your processing call should be printing some stats as it runs, so
paste some of that?



(You may have a study with a data type that makes files too frequently
because the data stream's frequency is set to zero.... its a bug that we
haven't gotten to ...)

**Comment by usanchez** [comment]
>Okay, sounds like you are uploading more data than you can process, hooray! Do some calls of `FileToProcess.objects.count()`. If your python-fu is up to it make a loop that prints that value every..... 10 seconds? Actually, your processing call should be printing some stats as it runs, so paste some of that?

I pressed Ctrl+c because I wanted to copy the stats, but it stopped the process instead. So I had to restart the process again, I hope that I didn't crew it up 😅 

The loops tells me that is processing approx. 500 files per 10 secs and that there are 8k files remaining. I'll edit with the updates.

> (You may have a study with a data type that makes files too frequently because the data stream's frequency is set to zero.... its a bug that we haven't gotten to ...)

I think that all frequency values are non-zero, but... I think that I had this issue since May, and didn't realise it until this week, so... Yeah

**Comment by biblicabeebli** [comment]
Fantastic.

Bug: somehow some data got uploaded that looks like it was either corrupted or decryption/encryption failed.  That string of gibberish is expected to be a numerical text value, buuuuut it isn't.

There is more debugging into the apps, it would be helpful if we could at the very least get the os that this user is running on

Fix:
You will need to identify the file that is failing. It LOOKs like `akjk778i` is the participant causing this... guess I've forgotten the exact nature of those print statements...

You will have to identify that file and remove it from the FileToProcess database table.  Lots of ways to do that, need help or do you think you can work that out?  (literally ending my commute right now)

**Comment by usanchez** [comment]
I think that I should open a new issue for this new bug, the processing seems to work fine.

I was thinking about removing that row from the table, is that a correct way to do it?

Then, I just uncomment the cronjobs and that's it right?

**Comment by biblicabeebli** [comment]
could you tell me the file name that was corrupted and check whether that
user is running on android or on ios?  Don't know how complex your setup
with studies is, do you need the python shell command line to get that data
point or can you work it out / get it from the website?

**Comment by usanchez** [comment]
> could you tell me the file name that was corrupted and check whether that user is running on android or on ios? Don't know how complex your setup with studies is, do you need the python shell command line to get that data point or can you work it out / get it from the website? 

This is the filename:
KN27Sbd8hbemieSBxgTBFPIw/dv8jxbt2/accel/1562076699521.csv

So it's accelerator data for participant dv8jxbt2, who is using an Android device. I think I need some help getting just _that_ single file unencrypted. And what do you mean by study setup complexity in this case? 

> I think you are all set, yes just reenable the cron job on the manager server.

One question related to this. I haven't activated the cron jobs yet. In the database, in the table filetoprocess, I only have the problematic file. But shouldn't be on this table also the rest of the files that have been uploaded to S3 that have not been processed yet?

> Recommendation: python 3 is nearly ready. Recommend you upgrade to it when it is merged into master/production. Python 2 is totally out-of-support (including security updates) on January 1st. Additionally the python 3 cluster allows for reduced server costs because the manager server will now also process data, so if you have low load you don't need that extra worker server. Finally, would you mind reviewing the instructions that were just put up on the wiki for doing a python 3 cluster upgrade and let me know if there is anything unclear in those directions? If so please file it as a new issue. 

Ok. I have reviewed the steps, and there is something not clear for me. I'll post the issue.

> Thanks for your patience in debugging this, sorry about the hassle and extra steps involved.

You were the one with the patience to help me debug it, it is a pleasure for me to help back in anything you could need.

> Sentry does have a free tier, it will make your life easier and your email a littlw less spammy.

Already used all the messages xD Seems easy to set up a Sentry server, I'm gonna give it a try.

**Comment by biblicabeebli** [comment]
>  And what do you mean by study setup complexity in this case?

I just meant if you had a lot of users then I could write you a command to just get the info instead of having you go through the UI searching for a user.

> So it's accelerator data for participant dv8jxbt2, who is using an Android device. I think I need some help getting just that single file unencrypted.

```
p = Participant.objects.get(patient_id="dv8jxbt2")
data = s3_retrieve(THAT_FILE_PATH, p.study.object_id)
```
You might  need to throw in ```raw_path=True```
DON'T POST THE CONTENTS.  That would probably break some kind of data privacy law.
Look for a line that is .... malformed?  It is supposed to just be a csv file.  ```data.splitlines()``` is your friend...
If there is an obviously bad line just remove it.  Do update on what the error is.  Check if it is missing the csv header line, that was possible on old (_ancient_) versions of the app for that to be missing, and there were iOS builds that did that too.  (This was well before this was open sourced, shouldn't be occurring.)
See if you can work out how to use ```s3_upload``` to upload + reencrypt it, if you fix it then you can go ahead and try to run processing again.

>  KN27Sbd8hbemieSBxgTBFPIw/dv8jxbt2/accel/1562076699521.csv

Great.  That tells me the data stream and I now have an OS, I can now try and look to determine what has occurred.

> Ok. I have reviewed the steps, and there is something not clear for me. I don't know if it requires a new issue, but I am missing is what to do when I update the environment, do I need to change to the Python3 branch (or production branch in the future)?

Standard ```eb deploy``` operation from the appropriate branch.  You should wait until this branch has been merged into master/production, and you should always synchronize branches between the processing and elastic beanstalk servers.

> Related to that, do you have a date for the availability of the Python 3 version of the platform? I'll try the migration as soon as it is available and tell you about any issues.

I'm trying to do it _right now_.  There was an office move and I just got back to development.  Issue #73 had a report of a full stack deploy error, I just fixed it, that was the last major task that I had not tested.  I need to run a few tests on my staging setup, and then it will be merged to master and production.

> Already used all the messages xD Seems easy to set up a Sentry server, I'm gonna give it a try.

Yeah we run the sentry endpoint for the Android app...  I know that life...

**Comment by usanchez** [comment]
Sorry for taking me some time to continue with this thread. 

Today I was able to dedicate some time to it and I discovered that the problematic file is the only file that this user contains in S3, there are no other data from other sensors. However, I've seen that there is another user with the exact same id in another study.

The underlying cause of the error may be another one than just the problem with malformed data. Let me explain:

- Checking the path of the problematic file, you would deduce that the user dv8jxbt2, belongs to study KN27Sbd8hbemieSBxgTBFPIw (study id 11).
- If we check the path in S3, the file is there as expected.
- When I used the commands that you provided me, the content was encrypted so I couldn't see anything wrong, so I decided to check any other encrypted file to check if there was any visible difference. 
- To my surprise, I found that there was another user with the exact same patient_id in another study. 
- I went to check the database and I found that there is no user dv8jxbt2 in study id 11, but that that user is in study 9 (TyM4tK3RKkts0OnfDaLUGnHy).

So here I am in front of my screen, like, how?

If I discover anything else, I'll let you know ASAP.

Let me answer below to the rest of the concerns in your message:

> DON'T POST THE CONTENTS. That would probably break some kind of data privacy law.
> Look for a line that is .... malformed? It is supposed to just be a csv file. data.splitlines() is your friend...

I will just post whatever it seems it's wrong, not the full contents of the file, but it's encrypted, so I don't know what could be wrong. Is there a way to unencrypt this data I got with `data.splitlines()`?

> See if you can work out how to use s3_upload to upload + reencrypt it, if you fix it then you can go ahead and try to run processing again.

My guess was something like:

`
s3_upload(THAT_FILE_PATH, new_data_string_formatted_in_csv, p.study.object_id, raw_path=True)
`

Again, thanks for all your effort. I'm very happy to help even if it is just creating issues.

**Comment by usanchez** [comment]
I tried to retrieve the same file but using the `study.object_id` of the study that is in the path (`KN27Sbd8hbemieSBxgTBFPIw`) but the content is still encrypted and seems to be almost empty if I use this alternative `object_id`.

What should I do? Just remove this file and resume file processing jobs?

I think it would be cool to add a check to see if the information extracted from the path of the S3 file is correct or not. The path has the `object_id` of the study, the data type and the `patient_id`, but in this case, it is not correct. If this may sound interesting to you, I could try to write a piece of code for it.

**Comment by biblicabeebli** [comment]
> What should I do? Just remove this file and resume file processing jobs?
Yeah.  Do that.  This seems like a malformed file.

The `HandledError`: this is a file that is not formatted correctly, it occurs at upload.  In this case the csv row's key length either invalid.  This does not mean the file is corrupted, it does mean a single row has to be dropped. This is probably the result of a corrupted/interrupted write operation, this can happen when the device is out of storage.  It can be ignored.  (I've not seen that in a _texts log_ before... so that could mean something, I will try and look to see if there is some kind of possible injection that could be occurring...  This error is unrelated to the file csv formatting error you are experiencing in the data processing code.  (It sucks; sorry.)

The wrong file path: I don't know what has occurred there. The `s3_upload` function is where all data pushed to s3, and user names are globally unique.  I think that an inappropriate parameter was passed in at some point with the `study_object_id` parameter while you were fiddling.  Such a bug in the codebase would result in general system failure because file paths would be wrong.  Look for the presence of the public and private key files for the participant, those are created at user generation, and the user will only be registerable in that study.

`s3_retrieve` pulls in the file and decrypts it.  It does require the correct study be provided because each study has a unique key.  if you are seeing junk in the output of `s3_retrieve` and are passing in the correct `study_object_id` then that means the file is garbage.

**Comment by usanchez** [comment]
> The `HandledError`: this is a file that is not formatted correctly, it occurs at upload. In this case the csv row's key length either invalid. This does not mean the file is corrupted, it does mean a single row has to be dropped. This is probably the result of a corrupted/interrupted write operation, this can happen when the device is out of storage. It can be ignored. (I've not seen that in a _texts log_ before... so that could mean something, I will try and look to see if there is some kind of possible injection that could be occurring... This error is unrelated to the file csv formatting error you are experiencing in the data processing code. (It sucks; sorry.)

I have a lot of cases for this HandledError, the `error_log` is filled with these errors, e.g.:

```
[Sun Dec 01 21:01:02.044602 2019] [:error] [pid 20789] ===================
[Sun Dec 01 21:01:02.044613 2019] [:error] [pid 20789] 2019-12-01T21:01:02.044398
[Sun Dec 01 21:01:02.044617 2019] [:error] [pid 20789] o1oysjik; uk7lwho7_callLog_1575108084710.csv; There was an error in user decryption: AES key has bad length.
[Sun Dec 01 21:01:02.044621 2019] [:error] [pid 20789] ERROR:HandledError('There was an error in user decryption: AES key has bad length.',)
[Sun Dec 01 21:01:02.044623 2019] [:error] [pid 20789] Traceback (most recent call last):
[Sun Dec 01 21:01:02.044626 2019] [:error] [pid 20789]   File "/opt/python/current/app/api/mobile_api.py", line 105, in upload
[Sun Dec 01 21:01:02.044629 2019] [:error] [pid 20789]     uploaded_file = decrypt_device_file(patient_id, uploaded_file, client_private_key, user)
[Sun Dec 01 21:01:02.044631 2019] [:error] [pid 20789]   File "/opt/python/current/app/libs/encryption.py", line 221, in decrypt_device_file
[Sun Dec 01 21:01:02.044634 2019] [:error] [pid 20789]     raise HandledError(error_message)
[Sun Dec 01 21:01:02.044636 2019] [:error] [pid 20789] HandledError: There was an error in user decryption: AES key has bad length.
[Sun Dec 01 21:01:02.044638 2019] [:error] [pid 20789]
[Sun Dec 01 21:01:02.044641 2019] [:error] [pid 20789] ===================
[Sun Dec 01 21:01:02.044645 2019] [:error] [pid 20789]
[Sun Dec 01 21:01:02.280317 2019] [:error] [pid 20790] the following error was handled:
[Sun Dec 01 21:01:02.280512 2019] [:error] [pid 20790] ===================
[Sun Dec 01 21:01:02.280541 2019] [:error] [pid 20790] 2019-12-01T21:01:02.280377
[Sun Dec 01 21:01:02.280544 2019] [:error] [pid 20790] o1oysjik; uk7lwho7_powerState_1574663426463.csv; There was an error in user decryption: AES key has bad length.
[Sun Dec 01 21:01:02.280548 2019] [:error] [pid 20790] ERROR:HandledError('There was an error in user decryption: AES key has bad length.',)
[Sun Dec 01 21:01:02.280551 2019] [:error] [pid 20790] Traceback (most recent call last):
[Sun Dec 01 21:01:02.280554 2019] [:error] [pid 20790]   File "/opt/python/current/app/api/mobile_api.py", line 105, in upload
[Sun Dec 01 21:01:02.280557 2019] [:error] [pid 20790]     uploaded_file = decrypt_device_file(patient_id, uploaded_file, client_private_key, user)
[Sun Dec 01 21:01:02.280559 2019] [:error] [pid 20790]   File "/opt/python/current/app/libs/encryption.py", line 221, in decrypt_device_file
[Sun Dec 01 21:01:02.280562 2019] [:error] [pid 20790]     raise HandledError(error_message)
[Sun Dec 01 21:01:02.280564 2019] [:error] [pid 20790] HandledError: There was an error in user decryption: AES key has bad length.
[Sun Dec 01 21:01:02.280567 2019] [:error] [pid 20790]
[Sun Dec 01 21:01:02.280569 2019] [:error] [pid 20790] ===================
[Sun Dec 01 21:01:02.280574 2019] [:error] [pid 20790]
[Sun Dec 01 21:01:04.172282 2019] [:error] [pid 20789] the following error was handled:
[Sun Dec 01 21:01:04.172442 2019] [:error] [pid 20789] ===================
[Sun Dec 01 21:01:04.172449 2019] [:error] [pid 20789] 2019-12-01T21:01:04.172321
[Sun Dec 01 21:01:04.172452 2019] [:error] [pid 20789] o1oysjik; uk7lwho7_callLog_1565792837443.csv; There was an error in user decryption: AES key has bad length.
[Sun Dec 01 21:01:04.172456 2019] [:error] [pid 20789] ERROR:HandledError('There was an error in user decryption: AES key has bad length.',)
[Sun Dec 01 21:01:04.172459 2019] [:error] [pid 20789] Traceback (most recent call last):
[Sun Dec 01 21:01:04.172461 2019] [:error] [pid 20789]   File "/opt/python/current/app/api/mobile_api.py", line 105, in upload
[Sun Dec 01 21:01:04.172465 2019] [:error] [pid 20789]     uploaded_file = decrypt_device_file(patient_id, uploaded_file, client_private_key, user)
[Sun Dec 01 21:01:04.172467 2019] [:error] [pid 20789]   File "/opt/python/current/app/libs/encryption.py", line 221, in decrypt_device_file
[Sun Dec 01 21:01:04.172470 2019] [:error] [pid 20789]     raise HandledError(error_message)
[Sun Dec 01 21:01:04.172472 2019] [:error] [pid 20789] HandledError: There was an error in user decryption: AES key has bad length.
[Sun Dec 01 21:01:04.172489 2019] [:error] [pid 20789]
[Sun Dec 01 21:01:04.172492 2019] [:error] [pid 20789] ===================
[Sun Dec 01 21:01:04.172496 2019] [:error] [pid 20789]
[Sun Dec 01 21:01:06.132737 2019] [:error] [pid 20789] the following error was handled:
[Sun Dec 01 21:01:06.132921 2019] [:error] [pid 20789] ===================
[Sun Dec 01 21:01:06.132929 2019] [:error] [pid 20789] 2019-12-01T21:01:06.132794
[Sun Dec 01 21:01:06.132933 2019] [:error] [pid 20789] o1oysjik; uk7lwho7_logFile_1574875856449.csv; There was an error in user decryption: AES key has bad length.
[Sun Dec 01 21:01:06.132936 2019] [:error] [pid 20789] ERROR:HandledError('There was an error in user decryption: AES key has bad length.',)
[Sun Dec 01 21:01:06.132939 2019] [:error] [pid 20789] Traceback (most recent call last):
[Sun Dec 01 21:01:06.132942 2019] [:error] [pid 20789]   File "/opt/python/current/app/api/mobile_api.py", line 105, in upload
[Sun Dec 01 21:01:06.132944 2019] [:error] [pid 20789]     uploaded_file = decrypt_device_file(patient_id, uploaded_file, client_private_key, user)
[Sun Dec 01 21:01:06.132947 2019] [:error] [pid 20789]   File "/opt/python/current/app/libs/encryption.py", line 221, in decrypt_device_file
[Sun Dec 01 21:01:06.132950 2019] [:error] [pid 20789]     raise HandledError(error_message)
[Sun Dec 01 21:01:06.132952 2019] [:error] [pid 20789] HandledError: There was an error in user decryption: AES key has bad length.
[Sun Dec 01 21:01:06.132955 2019] [:error] [pid 20789]
[Sun Dec 01 21:01:06.132957 2019] [:error] [pid 20789] ===================

```

What seems strange for me is the following line:

> [Sun Dec 01 21:01:04.172452 2019] [:error] [pid 20789] **o1oysjik**; **uk7lwho7**_callLog_1565792837443.csv; There was an error in user decryption: AES key has bad length.

The strings in bold are two different users... and I've got many errors like this in `error_log`. However, in the latest lines, I've got errors when the two users are the same:


> [Mon Dec 02 13:45:31.694008 2019] [:error] [pid 11724] the following error was handled:
> [Mon Dec 02 13:45:31.694212 2019] [:error] [pid 11724] ===================
> [Mon Dec 02 13:45:31.694220 2019] [:error] [pid 11724] 2019-12-02T13:45:31.694076
> [Mon Dec 02 13:45:31.694223 2019] [:error] [pid 11724] **kotjn778**; **kotjn778**_callLog_1575154244967.csv; There was an error in user decryption: AES key has bad length.
> [Mon Dec 02 13:45:31.694226 2019] [:error] [pid 11724] ERROR:HandledError('There was an error in user decryption: AES key has bad length.',)
> [Mon Dec 02 13:45:31.694229 2019] [:error] [pid 11724] Traceback (most recent call last):
> [Mon Dec 02 13:45:31.694232 2019] [:error] [pid 11724]   File "/opt/python/current/app/api/mobile_api.py", line 105, in upload
> [Mon Dec 02 13:45:31.694235 2019] [:error] [pid 11724]     uploaded_file = decrypt_device_file(patient_id, uploaded_file, client_private_key, user)
> [Mon Dec 02 13:45:31.694237 2019] [:error] [pid 11724]   File "/opt/python/current/app/libs/encryption.py", line 221, in decrypt_device_file
> [Mon Dec 02 13:45:31.694240 2019] [:error] [pid 11724]     raise HandledError(error_message)
> [Mon Dec 02 13:45:31.694242 2019] [:error] [pid 11724] HandledError: There was an error in user decryption: AES key has bad length.
> [Mon Dec 02 13:45:31.694245 2019] [:error] [pid 11724]
> [Mon Dec 02 13:45:31.694247 2019] [:error] [pid 11724] ===================
> [Mon Dec 02 13:45:31.694250 2019] [:error] [pid 11724]
> [Mon Dec 02 13:45:38.475688 2019] [:error] [pid 11724] the following error was handled:
> [Mon Dec 02 13:45:38.475878 2019] [:error] [pid 11724] ===================
> [Mon Dec 02 13:45:38.475887 2019] [:error] [pid 11724] 2019-12-02T13:45:38.475752
> [Mon Dec 02 13:45:38.475890 2019] [:error] [pid 11724] **kotjn778**; **kotjn778**_bluetoothLog_1575120300404.csv; There was an error in user decryption: AES key has bad length.
> [Mon Dec 02 13:45:38.475893 2019] [:error] [pid 11724] ERROR:HandledError('There was an error in user decryption: AES key has bad length.',)
> [Mon Dec 02 13:45:38.475896 2019] [:error] [pid 11724] Traceback (most recent call last):
> [Mon Dec 02 13:45:38.475899 2019] [:error] [pid 11724]   File "/opt/python/current/app/api/mobile_api.py", line 105, in upload
> [Mon Dec 02 13:45:38.475902 2019] [:error] [pid 11724]     uploaded_file = decrypt_device_file(patient_id, uploaded_file, client_private_key, user)
> [Mon Dec 02 13:45:38.475904 2019] [:error] [pid 11724]   File "/opt/python/current/app/libs/encryption.py", line 221, in decrypt_device_file
> [Mon Dec 02 13:45:38.475907 2019] [:error] [pid 11724]     raise HandledError(error_message)
> [Mon Dec 02 13:45:38.475909 2019] [:error] [pid 11724] HandledError: There was an error in user decryption: AES key has bad length.
> [Mon Dec 02 13:45:38.475911 2019] [:error] [pid 11724]
> [Mon Dec 02 13:45:38.475913 2019] [:error] [pid 11724] ===================
> [Mon Dec 02 13:45:38.475917 2019] [:error] [pid 11724]
> [Mon Dec 02 13:45:39.500942 2019] [:error] [pid 11723] the following error was handled:
> [Mon Dec 02 13:45:39.501136 2019] [:error] [pid 11723] ===================
> [Mon Dec 02 13:45:39.501144 2019] [:error] [pid 11723] 2019-12-02T13:45:39.501006
> [Mon Dec 02 13:45:39.501147 2019] [:error] [pid 11723] **kotjn778**; **xe58yqgq**_bluetoothLog_1575058500084.csv; There was an error in user decryption: AES key has bad length.
> [Mon Dec 02 13:45:39.501151 2019] [:error] [pid 11723] ERROR:HandledError('There was an error in user decryption: AES key has bad length.',)
> [Mon Dec 02 13:45:39.501168 2019] [:error] [pid 11723] Traceback (most recent call last):
> [Mon Dec 02 13:45:39.501170 2019] [:error] [pid 11723]   File "/opt/python/current/app/api/mobile_api.py", line 105, in upload
> [Mon Dec 02 13:45:39.501173 2019] [:error] [pid 11723]     uploaded_file = decrypt_device_file(patient_id, uploaded_file, client_private_key, user)
> [Mon Dec 02 13:45:39.501176 2019] [:error] [pid 11723]   File "/opt/python/current/app/libs/encryption.py", line 221, in decrypt_device_file
> [Mon Dec 02 13:45:39.501178 2019] [:error] [pid 11723]     raise HandledError(error_message)
> [Mon Dec 02 13:45:39.501180 2019] [:error] [pid 11723] HandledError: There was an error in user decryption: AES key has bad length.
> [Mon Dec 02 13:45:39.501183 2019] [:error] [pid 11723]
> [Mon Dec 02 13:45:39.501185 2019] [:error] [pid 11723] ===================
> [Mon Dec 02 13:45:39.501188 2019] [:error] [pid 11723]

I've got at least 1,690 instances of these `HandledError`s in the latest `error_log`.

I want to download all data and load it in R to see & analyse the contents. Any recommendation on how to do this?

> The wrong file path: I don't know what has occurred there. The `s3_upload` function is where all data pushed to s3, and user names are globally unique. I think that an inappropriate parameter was passed in at some point with the `study_object_id` parameter while you were fiddling. Such a bug in the codebase would result in general system failure because file paths would be wrong. Look for the presence of the public and private key files for the participant, those are created at user generation, and the user will only be registerable in that study.

The files are first decrypted with the private key of the participant to then be encrypted with the key of the study right? And what if the participant data was originally decrypted with an incorrect key? How can I check this? Is any such method available?

> `s3_retrieve` pulls in the file and decrypts it. It does require the correct study be provided because each study has a unique key. if you are seeing junk in the output of `s3_retrieve` and are passing in the correct `study_object_id` then that means the file is garbage.

So, I'll just delete the file and resume the processing of the rest of the files.

Thank you very much!

**Comment by biblicabeebli** [comment]
> I've got at least 1,690 instances of these HandledErrors in the latest error_log.
Oh goodie.

Private and Public keys: only the data uploaded by the participants uses asymmetric encryption.  (Architecture is this way so that data cannot be decrypted on the device after being written, only the server has the corresponding private key so only the server can decrypt it.)  Uploaded data is decrypted and then reencrypted using a per-study symmetric key for storage on s3.

**Comment by biblicabeebli** [comment]
Actually, could you post those errors? (not sure if you mean it is in an apache log or from the database, though your pasted content is clearly an apache log.)  I want to grep through that for patterns and similarities.

Also, where are  you? I'm wondering if your ios devices are from a non-US app store and therefore might be out of date.

**Comment by usanchez** [comment]
I'll collect all the "HandledError"s form the error logs and upload them somewhere so you can access them to analyse them. 

All our users are Android users, from Spain. We couldn't make the iOS app work, I think, due to incompatibilitites related to the app in beiwe-ios and the newest XCode.

**Comment by biblicabeebli** [comment]
Ah, sorry to hear. Apple requires that apps get ported to the newest
version of Swift about once a year, but only if you submit a new version.
You probably got unlucky and were caught in a stretch where we did not need
to update the app over in the US.  To my knowledge it is currently
compilable on the most recent versions of swift.

**Comment by usanchez** [comment]
Don't worry, I just took your work and tried making it work, can't complain if it doesn't. I'll try to make it work again this week and I'll post any issues to `beiwe-ios` if needed. 

Find attached the files for the `HandledErrors`. I included the data in an `RDS` file and an `XLSX` file, and also the code (in `R`) that I used to structure the error logs.
[error_log_analysis.zip](https://github.com/onnela-lab/beiwe-backend/files/3941371/error_log_analysis.zip)

**Comment by usanchez** [comment]
Before resuming the file processing, I analysed the filetoprocess table to see if there was any inconsistency in the data, for example, to check if the participant of the file to process is the same as the participant "code" that is stated in the S3 path that is contained in another column.

To understand the content of the table better, I have created some info codes for the name of the columns:

- _p_: refers to participant information.
- _s_: refers to study information.
- _s3_: refers to data extracted from the s3 path column (e.g. we can extract the study_object id, the participant_code, data type, etc.)

If more than one code appears in one column name, the order stablishes the relationship order, e.g.:

`ftp_s3_p_study_id`: from the data extracted from the `s3` filepath column, we have obtained the `patient` info (patient random code in this case), and with this patient info, we have obtained the id of the `study` that this participant belongs to.

You will find 2 sets of columns that contain the information about the consistency in this filetoprocess table. It compares the data in the column tables with the data extracted from the S3 file path.

For the study information consistency, I compared the `ftp_p_study_id`, `ftp_s_id`, `ftp_s3_p_study_id` and `ftp_s3_s_id` columns and for the participant information consistency I compared the `ftp_p_id` and `ftp_s3_p_id` columns.

Attached you can find the source code to extract all data used here, an `RDS` file to load into `R` and an `XLSX` file to load it with any other programming language or to visually inspect it.

[test_filetoprocess.zip](https://github.com/onnela-lab/beiwe-backend/files/3950300/test_filetoprocess.zip)

**Comment by usanchez** [comment]
So, as an update, after upgrading the platform and launching the new manager-worker server. All the files were processed except 191 files. I'll try to see if there are any logged errors and post them here.

**Comment by biblicabeebli** [comment]
Good news: I just finished (I think, still on development branch) rewriting the way we dispatch tasks and FileProcessLock will be removed from the codebase completely.

I'm going to close this issue, but future debuggers should read through this megathread as it will probably always be relevant.
